From 1ffdfbb3d55ec7a7a1b1a84e43a80e0ae96049db Mon Sep 17 00:00:00 2001
From: Jakub Jelinek <jakub@redhat.com>
Date: Fri, 5 Sep 2025 10:59:42 +0200
Subject: [PATCH] testsuite, powerpc, v2: Fix vsx-vectorize-* after alignment
 peeling [PR118567]

On Tue, Jul 01, 2025 at 02:50:40PM -0500, Segher Boessenkool wrote:
> No tests become good tests without effort.  And tests that are not good
> tests require constant maintenance!

Here are two patches, either just the first one or both can be used
and both were tested on powerpc64le-linux.

The second one adds further 8 tests, which are dg-do run which #include
the former tests, don't do any dump tests and just define the checking/main
for those.

2025-09-05  Jakub Jelinek  <jakub@redhat.com>

	PR testsuite/118567
	* gcc.target/powerpc/vsx-vectorize-9.c: New test.
	* gcc.target/powerpc/vsx-vectorize-10.c: New test.
	* gcc.target/powerpc/vsx-vectorize-11.c: New test.
	* gcc.target/powerpc/vsx-vectorize-12.c: New test.
	* gcc.target/powerpc/vsx-vectorize-13.c: New test.
	* gcc.target/powerpc/vsx-vectorize-14.c: New test.
	* gcc.target/powerpc/vsx-vectorize-15.c: New test.
	* gcc.target/powerpc/vsx-vectorize-16.c: New test.

(cherry picked from commit 8d8b6249d83609e56752ee51d2686b2f5fb062e6)
---
 .../gcc.target/powerpc/vsx-vectorize-10.c     | 39 +++++++++++++++++++
 .../gcc.target/powerpc/vsx-vectorize-11.c     | 36 +++++++++++++++++
 .../gcc.target/powerpc/vsx-vectorize-12.c     | 36 +++++++++++++++++
 .../gcc.target/powerpc/vsx-vectorize-13.c     | 36 +++++++++++++++++
 .../gcc.target/powerpc/vsx-vectorize-14.c     | 39 +++++++++++++++++++
 .../gcc.target/powerpc/vsx-vectorize-15.c     | 39 +++++++++++++++++++
 .../gcc.target/powerpc/vsx-vectorize-16.c     | 39 +++++++++++++++++++
 .../gcc.target/powerpc/vsx-vectorize-9.c      | 24 ++++++++++++
 8 files changed, 288 insertions(+)
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-10.c
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-11.c
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-12.c
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-13.c
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-14.c
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-15.c
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-16.c
 create mode 100644 gcc/testsuite/gcc.target/powerpc/vsx-vectorize-9.c

diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-10.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-10.c
new file mode 100644
index 000000000000..4a712fb21cd9
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-10.c
@@ -0,0 +1,39 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-95.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-2.c"
+
+__attribute__ ((noinline))
+void bar (float *pd, float *pa, float *pb, float *pc) 
+{
+  int i;
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (pa[i] != (pb[i] * pc[i]))
+	abort ();
+      if (pd[i] != 5.0)
+	abort ();
+    }
+
+  return;
+}
+
+int main (void)
+{
+  int i;
+  float a[N] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  float d[N+1] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  float b[N] = {0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57};
+  float c[N] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19};
+
+  main1 (N,&d[1],a,b,c);
+  main1 (N-2,&d[1],a,b,c);
+
+  return 0;
+}
diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-11.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-11.c
new file mode 100644
index 000000000000..55f909bcc7b4
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-11.c
@@ -0,0 +1,36 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-95.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-3.c"
+
+__attribute__ ((noinline))
+void bar (short *pa, short *pb, short *pc) 
+{
+  int i;
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (pa[i] != (pb[i] * pc[i]))
+	abort ();
+    }
+
+  return;
+}
+
+int main (void)
+{
+  int i;
+  short a[N] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  short b[N] = {0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57};
+  short c[N] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19};
+
+  main1 (N,a,b,c);
+  main1 (N-2,a,b,c);
+
+  return 0;
+}
diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-12.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-12.c
new file mode 100644
index 000000000000..5aa86c9455e2
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-12.c
@@ -0,0 +1,36 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-95.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-4.c"
+
+__attribute__ ((noinline))
+void bar (double *pa, double *pb, double *pc) 
+{
+  int i;
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (pa[i] != (pb[i] * pc[i]))
+	abort ();
+    }
+
+  return;
+}
+
+int main (void)
+{
+  int i;
+  double a[N] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  double b[N] = {0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57};
+  double c[N] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19};
+
+  main1 (N,a,b,c);
+  main1 (N-2,a,b,c);
+
+  return 0;
+}
diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-13.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-13.c
new file mode 100644
index 000000000000..be179a4b5966
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-13.c
@@ -0,0 +1,36 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-95.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-5.c"
+
+__attribute__ ((noinline))
+void bar (char *pa, char *pb, char *pc) 
+{
+  int i;
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (pa[i] != (pb[i] + pc[i]))
+	abort ();
+    }
+
+  return;
+}
+
+int main (void)
+{
+  int i;
+  char a[N] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  char b[N] = {0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57};
+  char c[N] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19};
+
+  main1 (N,a,b,c);
+  main1 (N-2,a,b,c);
+
+  return 0;
+}
diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-14.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-14.c
new file mode 100644
index 000000000000..18ce41ee2d54
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-14.c
@@ -0,0 +1,39 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-95.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-6.c"
+
+__attribute__ ((noinline))
+void bar (double *pd, double *pa, double *pb, double *pc) 
+{
+  int i;
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (pa[i] != (pb[i] * pc[i]))
+	abort ();
+      if (pd[i] != 5.0)
+	abort ();
+    }
+
+  return;
+}
+
+int main (void)
+{
+  int i;
+  double a[N] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  double d[N+1] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  double b[N] = {0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57};
+  double c[N] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19};
+
+  main1 (N,&d[1],a,b,c);
+  main1 (N-2,&d[1],a,b,c);
+
+  return 0;
+}
diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-15.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-15.c
new file mode 100644
index 000000000000..5549a4df2d47
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-15.c
@@ -0,0 +1,39 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-95.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-7.c"
+
+__attribute__ ((noinline))
+void bar (int *pd, int *pa, int *pb, int *pc) 
+{
+  int i;
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (pa[i] != (pb[i] * pc[i]))
+	abort ();
+      if (pd[i] != 5.0)
+	abort ();
+    }
+
+  return;
+}
+
+int main (void)
+{
+  int i;
+  int a[N] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  int d[N+1] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  int b[N] = {0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57};
+  int c[N] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19};
+
+  main1 (N,&d[1],a,b,c);
+  main1 (N-2,&d[1],a,b,c);
+
+  return 0;
+}
diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-16.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-16.c
new file mode 100644
index 000000000000..82fc07b34096
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-16.c
@@ -0,0 +1,39 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-95.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-8.c"
+
+__attribute__ ((noinline))
+void bar (short *pd, short *pa, short *pb, short *pc) 
+{
+  int i;
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (pa[i] != (pb[i] * pc[i]))
+	abort ();
+      if (pd[i] != 5.0)
+	abort ();
+    }
+
+  return;
+}
+
+int main (void)
+{
+  int i;
+  short a[N] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  short d[N+1] __attribute__ ((__aligned__(__BIGGEST_ALIGNMENT__)));
+  short b[N] = {0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57};
+  short c[N] = {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19};
+
+  main1 (N,&d[1],a,b,c);
+  main1 (N-2,&d[1],a,b,c);
+
+  return 0;
+}
diff --git a/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-9.c b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-9.c
new file mode 100644
index 000000000000..9e87edcdd25c
--- /dev/null
+++ b/gcc/testsuite/gcc.target/powerpc/vsx-vectorize-9.c
@@ -0,0 +1,24 @@
+/* { dg-do run } */
+/* { dg-options "-O2 -ftree-vectorize -fno-tree-loop-distribute-patterns -fno-vect-cost-model -fdump-tree-vect-details" } */
+
+/* Taken from vect/vect-align-1.c.  */
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include "vsx-vectorize-1.c"
+
+int main (void)
+{
+  int i;
+  struct foo *p = malloc (2*sizeof (struct foo));
+  
+  main1 (p);
+
+  /* check results:  */
+  for (i = 0; i < N; i++)
+    {
+      if (p->y[i] != x[i])
+	abort ();
+    }
+  return 0;
+}
-- 
2.43.7


